[Clang] Remove 3-element vector load and store special handling

shiltian opened the pull request:
Clang uses a long-time special handling of the case where  3 element vector loads and stores are performed as 4 element, and then a shufflevector is used to extract the used elements. Odd sized vector codegen should now work reasonably well.

This patch removes the compiler argument `-fpreserve-vec3-type` and adds a target hook to determine if the special handling of vector type is needed.

shiltian commented:
* **#104661** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/104661?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @shiltian and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

topperc commented on clang/test/CodeGen/alignment.c:
```diff
@@ -45,7 +45,7 @@ void test3(packedfloat3 *p) {
 }
 // CHECK: @test3(
-// CHECK: store <4 x float> {{.*}}, align 4
+// CHECK: store <3 x float> {{.*}}, align 4
```
Is the backend able to lower this a a single store? Or will it get split into 2 or more stores? As far as I know the X86 backend will split it into 2 stores, a 64 bit and a 32 bit.

arsenm replied:
According to the langref the backend can widen the store up to the alignment, if not using sanitizers. It just doesn't today

efriedma-quic replied:
Even if we can prove it doesn't trap, it's still not legal to store to the trailing bytes; they could contain other data.  So we have to split the store.

arsenm replied:
I guess this has to be a target option then

shiltian replied:
So what is the conclusion here? Do we want to have a custom lowering for X86?

efriedma-quic replied:
On targets with SIMD vectors (basically any modern CPU target), a 12-byte store is going to be slower than a 16-byte store, so we don't want to generate a 12-byte store.  So clang has an ABI rule that says we can generate a 16-byte store instead.

If you have some target that actually wants 12-byte stores for some reason, it needs to be a target-specific ABI rule in clang: on that target, we just directly generate 12-byte stores, instead of 16-byte stores.

shiltian replied:
Hmm, that sounds essentially what the option that was removed in this PR is doing.

shiltian replied:
Does it make sense to move the logic to back end? If a target doesn't support odd-byte load/store, we legalize it.

efriedma-quic replied:
> that sounds essentially what the option that was removed in this PR is doing

Yes, exactly.

> Does it make sense to move the logic to back end

The way LLVM IR is defined, a `<3 x float>` stores 12 bytes, and we can't really change that.

arsenm replied:
The ideal control would be more specific than this. We have legal 12-byte load/store, but not 3/6 or other odd sizes

shiltian replied:
Then I guess this can be more general: add a virtual target specific function to return the next legal size given N. If the two numbers don't match, we can extend the handling here to support "arbitrary" N.

arsenm replied:
I'd expect this to be in terms of type, not size but yes.

shiltian replied:
I took another look and this is not only a target option. For example, if it is hlsl, no matter what target it is (spir-v or dxil), vec3 is preserved. I tried to add a function `TargetInfo::getOptimalVectorNumElements(const llvm::FixedVectorType *)`, but it can't cover the language related stuff. If we add an extra argument, like `bool IsHLSL)`, it is too ugly.

shiltian replied:
Eventually we might do it like:

```
bool preserveVec3(Triple TT, bool IsHLSL) {
  if (IsHLSL)
    return true;
  if (TT.isAMDGPU())
    return true;
  return false;
}
```

arsenm replied:
I don't follow. The language merely states if the type has the padding bytes. It doesn't mandate you optimize down to the used elements

arsenm replied:
Don't see what's so ugly. The target hook (which should be a function of the type) would be aware of the language options

shiltian replied:
`unsigned getOptimalVectorNumElements(const llvm::FixedVectorType *, const LangOptions &)`?

arsenm replied:
I would return the VectorType to use instead of doing it in terms of elements. Maybe the input should be the QualType?

shiltian commented on clang/include/clang/Driver/Options.td:
```diff
@@ -8026,10 +8026,6 @@ def fhlsl_strict_availability : Flag<["-"], "fhlsl-strict-availability">,
-def fpreserve_vec3_type : Flag<["-"], "fpreserve-vec3-type">,
-  HelpText<"Preserve 3-component vector type">,
-  MarshallingInfoFlag<CodeGenOpts<"PreserveVec3Type">>,
-  ImpliedByAnyOf<[hlsl.KeyPath]>;
```
@arsenm Here.

shiltian commented on clang/include/clang/Basic/TargetInfo.h:
```diff
@@ -1846,6 +1846,16 @@ class TargetInfo : public TransferrableTargetInfo,
+  virtual llvm::FixedVectorType *
+  getOptimalVectorType(llvm::FixedVectorType *T, const LangOptions &Opt) const {
+    if (!Opt.HLSL && T->getNumElements() == 3)
+      return llvm::FixedVectorType::get(T->getElementType(), 4);
```
The logic here might look weird, but it preserves the current behavior prior to this PR.

arsenm replied:
I think it would be better to keep this in a LangOptions field, regardless of whether we keep the flag or not.

shiltian replied:
can you explain more? I didn't follow.

arsenm replied:
LangOpts.PreserveVec3 or something. I don't want LangOpts.LanguageX. || LangOpts.LanguageY checks littered around, those should exist only in some LangOpts wrapper

shiltian commented on clang/lib/Basic/CMakeLists.txt:
```diff
@@ -1,4 +1,5 @@
 set(LLVM_LINK_COMPONENTS
+  Core
```
This is needed because `TargetInfo.cpp` now needs LLVM types.

topperc replied:
Does this create additional dependencies on clang based tools that don't need to use LLVM IR? For example, clang-format?

arsenm replied:
This shouldn't be necessary. I think this code is in the wrong place, I would expect it to go in lib/CodeGen/Targets, not lib/Basic

shiltian replied:
> This shouldn't be necessary. I think this code is in the wrong place, I would expect it to go in lib/CodeGen/Targets, not lib/Basic

And then to have clang depend on LLVM's CodeGen component? If that's the case, I don't think that is correct. That is even worse.

> Does this create additional dependencies on clang based tools that don't need to use LLVM IR? For example, clang-format?

That's a good question. I think it does. Do you have suggestions on how to design the interface?

arsenm replied:
No, this is clang lib/CodeGen which has nothing to do with actual codegen. Someone really needs to just go ahead with the rename to IRGen

topperc replied:
> No, this is clang lib/CodeGen which has nothing to do with actual codegen. Someone really needs to just go ahead with the rename to IRGen

Core is LLVM's lib/IR isn't it? This is making clang lib/Basic depend on llvm/lib/IR. It can't be making Basic depend on clang's lib/CodeGen, that would be a circular dependency.

arsenm replied:
This should be defined here https://github.com/llvm/llvm-project/blob/main/clang/lib/CodeGen/Targets/AMDGPU.cpp

shiltian commented:
gentle ping

topperc commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -2003,20 +2003,19 @@ llvm::Value *CodeGenFunction::EmitLoadOfScalar(Address Addr, bool Volatile,
+      unsigned OldNumElements = VTy->getNumElements();
+      SmallVector<int, 4> Mask(OldNumElements);
+      for (unsigned I = 0; I < OldNumElements; ++I)
+        Mask[I] = I;
```
Can we use std::iota to fill the vector?

topperc commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -2146,21 +2145,21 @@ void CodeGenFunction::EmitStoreOfScalar(llvm::Value *Value, Address Addr,
+      if (!ClangVecTy->isExtVectorBoolType() && VecTy != NewVecTy) {
+        SmallVector<int, 4> Mask(NewVecTy->getNumElements(), -1);
+        for (unsigned I = 0; I < VecTy->getNumElements(); ++I)
+          Mask[I] = I;
```
std::iota?

shiltian commented:
ping @arsenm @topperc

arsenm commented on clang/lib/CodeGen/ABIInfo.h:
```diff
@@ -123,6 +124,12 @@ class ABIInfo {
+  /// example, on certain targets, a vector with 3 elements might be promoted to
+  /// one with 4 elements to improve performance.
+  virtual llvm::FixedVectorType *
+  getOptimalVectorType(llvm::FixedVectorType *T, const LangOptions &Opt) const;
```
VectorMemoryType?

arsenm replied:
should specify it's in in memory type in the name and comment

arsenm commented on clang/lib/CodeGen/Targets/AMDGPU.cpp:
```diff
@@ -52,6 +52,12 @@ class AMDGPUABIInfo final : public DefaultABIInfo {
+  llvm::FixedVectorType *
+  getOptimalVectorType(llvm::FixedVectorType *T,
+                       const LangOptions &) const override {
+    return T;
```
We only want to preserve this for the 96-bit cases where we have legal instructions

arsenm replied:
With a direct instruction, yes. We can't widen the stores in the other odd cases so we need the hint that it's OK by using the wider type

arsenm commented on clang/test/CodeGenOpenCL/preserve_vec3.cl:
```diff
@@ -1,77 +0,0 @@
-// RUN: %clang_cc1 %s -emit-llvm -o - -triple spir-unknown-unknown -fpreserve-vec3-type | FileCheck %s
-
-typedef char char3 __attribute__((ext_vector_type(3)));
```
Shouldn't delete the test

shiltian replied:
The point of the test is for preserving vec3, as the name of the file indicates, which is no longer the case. Why do we want to keep it?

arsenm replied:
I'm not sure we have adequate coverage of the 3 element vector types elsewhere. You could remove the flag

arsenm commented on clang/test/CodeGenOpenCL/amdgpu-alignment.cl:
```diff
@@ -94,43 +94,43 @@ typedef double __attribute__((ext_vector_type(16))) double16;
 // CHECK: store volatile i32 0, ptr addrspace(3) @local_memory_alignment_global.lds_i32, align 4
 // CHECK: store volatile <2 x i32> zeroinitializer, ptr addrspace(3) @local_memory_alignment_global.lds_v2i32, align 8
-// CHECK: store volatile <4 x i32> <i32 0, i32 0, i32 0, i32 undef>, ptr addrspace(3) @local_memory_alignment_global.lds_v3i32, align 16
+// CHECK: store volatile <3 x i32> zeroinitializer, ptr addrspace(3) @local_memory_alignment_global.lds_v3i32, align 16
```
Need to have some 32-bit and 64-bit pointer vector tests, I don't think this test covers them

arsenm replied:
`error: invalid vector element type '__private char *'`

I guess it doesn't let you use pointers for no real reason

arsenm commented on clang/lib/CodeGen/Targets/AMDGPU.cpp:
```diff
@@ -52,6 +52,14 @@ class AMDGPUABIInfo final : public DefaultABIInfo {
+  llvm::FixedVectorType *
+  getOptimalVectorType(llvm::FixedVectorType *T,
+                       const LangOptions &Opt) const override {
+    if (T->getNumElements() == 3 && T->getScalarSizeInBits() == 32)
```
Should use the datalayout to get the size

arsenm replied:
Comment why. Also should add a fixme that we should check subtarget feature. Technically SI didn't have these

shiltian commented on clang/lib/CodeGen/Targets/AMDGPU.cpp:
```diff
@@ -52,6 +52,17 @@ class AMDGPUABIInfo final : public DefaultABIInfo {
+    // We have legal instructions for 96-bit so 3x32 can be supported.
+    // FIXME: This check should be a subtarget feature as technically SI doesn't
+    // support it.
+    if (T->getNumElements() == 3 && getDataLayout().getTypeSizeInBits(T) == 96)
```
@arsenm I wonder if we want to relax it to simply check size to be 96 to support cases like `3x32` (which is covered at the moment), `6x16` and `12x8`.

arsenm replied:
Yes, it should be any 96-bit access

shiltian replied:
It looks like they are preserved anyway. Clang only has special handling of 3-element vector.

shiltian commented:
bump bump

shiltian committed:
[Clang] Remove 3-element vector load and store special handling

Clang uses a long-time special handling of the case where  3 element vector loads
and stores are performed as 4 element, and then a shufflevector is used to extract
the used elements. Odd sized vector codegen should now work reasonably well.

This patch removes this special handling, as well as the compiler argument
`-fpreserve-vec3-type`.

arsenm commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -2145,21 +2144,21 @@ void CodeGenFunction::EmitStoreOfScalar(llvm::Value *Value, Address Addr,
+      auto *NewVecTy =
+          CGM.getABIInfo().getOptimalVectorMemoryType(VecTy, getLangOpts());
+      if (!ClangVecTy->isExtVectorBoolType() && VecTy != NewVecTy) {
+        SmallVector<int, 4> Mask(NewVecTy->getNumElements(), -1);
```
```suggestion
        SmallVector<int, 16> Mask(NewVecTy->getNumElements(), -1);
```

arsenm commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -2002,20 +2002,19 @@ llvm::Value *CodeGenFunction::EmitLoadOfScalar(Address Addr, bool Volatile,
+      Address Cast = Addr.withElementType(NewVecTy);
+      llvm::Value *V = Builder.CreateLoad(Cast, Volatile, "loadVecN");
+      unsigned OldNumElements = VTy->getNumElements();
+      SmallVector<int, 4> Mask(OldNumElements);
```
```suggestion
      SmallVector<int, 16> Mask(OldNumElements);
```

shiltian committed:
Update clang/lib/CodeGen/CGExpr.cpp

Co-authored-by: Matt Arsenault <Matthew.Arsenault@amd.com>

shiltian committed:
Update clang/lib/CodeGen/CGExpr.cpp

Co-authored-by: Matt Arsenault <Matthew.Arsenault@amd.com>

shiltian commented:
The compile error was resolved by 59dffce8c80eb9cefc96b8d3fe55473edfee9c4c.
